crates/autopilot/src/boundary/events/settlement.rs

```diff
@@ -11,33 +11,48 @@ impl_event_retrieving! {
 
 pub struct Indexer {
     db: Postgres,
+    settlement_updater: OnSettlementEventUpdater,
 }
 
 impl Indexer {
-    pub fn new(db: Postgres) -> Self {
-        Self { db }
+    pub fn new(db: Postgres, settlement_updater: OnSettlementEventUpdater) -> Self {
+        Self {
+            db,
+            settlement_updater,
+        }
     }
 }
 
 #[async_trait::async_trait]
 impl EventStoring<contracts::gpv2_settlement::Event> for Indexer {
     async fn last_event_block(&self) -> Result<u64> {
-        let store: &dyn EventStoring<contracts::gpv2_settlement::Event> = &self.db;
-        store.last_event_block().await
+        let mut con = self.db.pool.acquire().await?;
+        crate::database::events::last_event_block(&mut con).await
     }
 
     async fn replace_events(
         &mut self,
         events: Vec<ethcontract::Event<contracts::gpv2_settlement::Event>>,
         range: RangeInclusive<u64>,
     ) -> Result<()> {
-        self.db.replace_events(events, range).await
+        let mut transaction = self.db.pool.begin().await?;
```

Comment:
The fact that we want everything to happen in an atomic transaction is what caused the discussion on #2335. Without leaking the "transaction abstraction" somehow in the domain, I find it hard to find an appropriate separation between infra (doing the updates) and domain (deciding what updates to do).

Reply:
Does it actually HAVE to happen in an atomic transaction? 
AFAIS both `replace` and `append` can be moved to `infra` and abstract the "transaction" object. The question remains if the OnSettlementEventUpdater has to be atomic. Again, we do a bunch of reads there (settlement without auction, auction external prices), but once we prepare the `AuctionData`, we do one write to db and it's done, which means that we can again send AuctionData to `infra` and handle transaction object inside.

Reply:
I see. We could probably define a `domain::Update` struct that contains a list of events and a `delete_from_block` field and pass this into an infra function which then applies all updates atomically.